demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -0,0 +1,732 @@
+ * 10 we get the maximum length of this header value as: 10 * 2 + 7 + 1.
+ */
+#define RANGE_VALUE_MAX_LENGTH                      ( 28 )
+
+/**
+ * @brief The time for the application task to wait to try again finding a free request from the pool of requests.
+ */
+#define GET_FREE_REQUEST_RETRY_WAIT_TIME_MS         ( 1000 )
+
+/**
+ * @brief HTTP standard header field "Range".
+ */
+#define RANGE_HEADER_FIELD          "Range"
+
+/**
+ * @brief HTTP standard header field "Content-Range"
+ */
+#define CONTENT_RANGE_HEADER_FIELD  "Content-Range"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief This structure defines data consumed and kept track of in the async callbacks per request.
+ */
+typedef struct _asyncDownloadData {
+    char rangeValueStr[RANGE_VALUE_MAX_LENGTH];     /* This string is generated outside of the callback context and used 
+                                                       to set the request Range header within the callback context. */
+    int reqNum;                 /* This is the current request number in the static pool of request memory. */
+    uint32_t currDownloaded;    /* The current number of bytes downloaded in this request. */
+    uint32_t numReqBytes;       /* The number of bytes we want to downloaded in this request. */
+} _asyncDownloadData_t;
+
+/*-----------------------------------------------------------*/  
+
+/**
+ * @brief Buffer to store parts of the response body.
+ * 
+ * Since all of the requests in this demo are on the same connection, only one async task at a time will be accessing 
+ * this buffer to retrieve data from the network to print. 
+ */
+static uint8_t _pRespBodyBuffer[IOT_DEMO_HTTPS_RESP_BODY_BUFFER_SIZE] = { 0 };
+
+/**
+ * @brief Semaphore use to signal that the demo is finished. The task that downloads the final increment of the file
+ * posts to this semaphore. 
+ */
+static IotSemaphore_t _fileFinishedSem = { 0 };
+
+/**
+ * @brief Array to keep track of HTTPS Client library request and response resources (buffers, handles, and configs) 
+ * in use. 
+ */
+static bool _pInUseRequests[IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS] = { 0 };
+
+/**
+ * @brief Mutex to protect retrieving and setting the requests in use. 
+ */
+static IotMutex_t _inUseRequestsMutex = { 0 };
+
+/**
+ * @brief The pool of HTTPS Client library request configurations. 
+ */
+static IotHttpsRequestInfo_t _pReqConfigs[IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS] = {0};
+
+/**
+ * @brief The pool of HTTPS Client library response configurations.
+ */
+static IotHttpsResponseInfo_t _pRespConfigs[IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS] = { 0 };
+
+/**
+ * @brief The pool of HTTPS Client library request handles. 
+ */
+static IotHttpsRequestHandle_t _pReqHandles[IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS] = { 0 };
+
+/**
+ * @brief The pool of HTTPS Client library response handles. 
+ */
+static IotHttpsResponseHandle_t _pRespHandles[IOT_HTTPS_DEMO_MAX_ASYNC_REQUESTS] = { 0 };
```

Comment:
Has all this been run on a MCU? I am concerned about the memory needs for having 3 async requests and associated buffers etc. .

Reply:
Yes this has been run on the ESP32 device. Will continue to test on other MCUs.

Reply:
All tests and demos pass on ST.